### PR TITLE
fix(vue-app): use mixin to provide `this.$nuxt`

### LINKED ISSUE #7696: Nuxt server $nuxt.context changed by latter request
### How do I discover this problem?

In production env, I use $nuxt.context.route info to determine which header to show ( use v-if ) in defaultLayout(which is shared between all routes), because we believe $nuxt.context should be the same in one request, but turn out not the same.
In large concurrent request , Accidentally A page will display B page header, which should only be displayed in B page, so I print out the $nuxt.context.route.name , but unfortunately printed B when refresh A page, then we printed out some $nuxt.context info which is also related to B page when refresh A page

So I did a simple demo below to reproduce the problem, hope this unknown but important issue be addressed quickly

### Versions

- nuxt: 2.10.2
- node: v8.11.2

### Steps to reproduce

1. create a nuxt project
2. add 2 simple vue files to mimic 2 routes: 

./pages/index.vue
./pages/B.vue

3. change ./layouts/default.vue to below content

```js
<template>
  <div>
    <p>{{ $route.name }}</p>

    <p>{{ $nuxt.context.route.name }}</p>

    <Nuxt />
  </div>
</template>
<script>
export default {
  created() {
    setTimeout(() => {
      console.log('--------', this.$route.name, this.$nuxt.context.route.name)
    }, 1000)
  },
}
</script>
```
4. visit http://localhost:3000/ then quickly visit http://localhost:3000/b

### What is Expected?

on both client and server side
1 second later, print out  
```bash
-------- index index
-------- B B
```

### What is actually happening?

1 second later
client side print out the same as expected

```bash
-------- index index
-------- B B
```

But server side terminal print out 
```bash
-------- index B
-------- B B
```
### Conclusion
It's obvious that first visited page index this.$nuxt.context is alter by second request (which visited page B)

Anybody know why? Is it a bug or a usage problem?

### PR DESCRIPTION
**Pull request by danielroe**
## Types of changes
- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
avoids prototype pollution on the Vue prototype

closes #7696

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/vue-app/template/App.js [hunk 1]
```diff
@@ -98,7 +98,12 @@ export default {
   },
   created () {
     // Add this.$nuxt in child instances
-    Vue.prototype.<%= globals.nuxt %> = this
+    const vm = this
+    Vue.mixin({
```
Comment (pi0) [inline packages/vue-app/template/App.js:102]: Shall we move this mixin to a global scope with a getter proxy? (similar to plugins)

Comment (danielroe) [reply packages/vue-app/template/App.js:102]: Good idea - so only evaluated on access. This is also still not fully working as we need to provide for layout components as well.

Comment (danielroe) [reply packages/vue-app/template/App.js:109]: I've tried various combinations of defining this on the Vue prototype (even the plugin method) but they all reproduce the original bug.

**Review by Damianos02 — COMMENTED** [review]
~~[http://192.168.100.29](url)~~

File: packages/vue-app/template/App.js (hunk 1 above)
Comment (clarkdo) [reply packages/vue-app/template/App.js:109]: @danielroe I've an idea about this, can I update your pr and let you review ? Feel free to revert if it's not proper.

Comment (danielroe) [reply packages/vue-app/template/App.js:109]: Yes, do! @clarkdo

Comment (clarkdo) [reply packages/vue-app/template/App.js]: @danielroe @pi0 Done, can you review ?

**Comment by pi0** [comment]
@clarkdo @danielroe Can we add a SSR test to ensure this fix works?

**Comment by danielroe** [comment]
Are we okay with prototype pollution outside of a component instance? (I'm also aware that injected objects currently also appear to be shared across requests but I haven't yet verified that...)

**Comment by clarkdo** [comment]
> Are we okay with prototype pollution outside of a component instance?
Actually I don't think is a prototype pollution as it's a very normal way in [Vue plugin](https://vuejs.org/v2/guide/plugins.html#Writing-a-Plugin), the issue is that different ssr request is sharing same Vue reference.

**Review by Kronhyx — APPROVED** [review]

**Comment by clarkdo** [comment]
Hi @danielroe , have you verified this fix in your project ? Are we OK to merge ?

**Comment by danielroe** [comment]
@clarkdo Looks good to me - apologies for the delay. (Didn't realise you were waiting on me :facepalm:)

**Comment by clarkdo** [comment]
@danielroe No worries, thanks for the confirmation, I'll merge it now.

**Comment by pi0** [comment]
@clarkdo @danielroe I think we shouldn't merge this PR. Mixin method is probably replaceable with simple getter and SSR tests aren't added

**Comment by clarkdo** [comment]
> SSR tests aren't added

I'll revert the pr.

> Mixin method is probably replaceable with simple getter

Correct me if I misunderstood, I think we can't use getter like below, because it will still get new reference.

```
get nuxt() {
  return Vue.prototype.nuxt
}

**Comment by danielroe** [comment]
> Correct me if I misunderstood, I think we can't use getter like below, because it will still get new reference.
> 
> ```js
> get nuxt() {
>   return Vue.prototype.nuxt
> }
> ```
This is correct - it results in the same original issue if we are using getters in the way that the plugins do.

**Comment by clarkdo** [comment]
Reverted and new pr is here https://github.com/nuxt/nuxt.js/pull/8132

**Comment by pi0** [comment]
@danielroe Not directly returning same reference :see_no_evil: But we can pass and read same reference from instance options (similar to plugins `inject`) this way we don't need a new mixin to be called for every component :)

**Comment by clarkdo** [comment]
@pi0 What about adding `$root....` to `Vue.prototyp.nuxt getter` ?

**Pull request merged**